provider/providertest/integration.go

```diff
@@ -0,0 +1,394 @@
+package providertest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/Masterminds/squirrel"
+	"github.com/cloudquery/cq-provider-sdk/cqproto"
+	"github.com/cloudquery/cq-provider-sdk/logging"
+	"github.com/cloudquery/cq-provider-sdk/provider"
+	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+	"github.com/go-test/deep"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/hashicorp/terraform-exec/tfexec"
+	"github.com/jackc/pgx/v4"
+	"github.com/tmccombs/hcl2json/convert"
+)
+
+const (
+	tfDir    = "./int_test/tf/"
+	tfOrigin = "./testData/"
+)
+
+type ResourceIntegrationTestData struct {
+	Table               *schema.Table
+	Config              interface{}
+	Resources           []string
+	Configure           func(logger hclog.Logger, data interface{}) (schema.ClientMeta, error)
+	Suffix              string
+	Prefix              string
+	VerificationBuilder func(res *ResourceIntegrationTestData) ResourceIntegrationVerification
+}
+
+// ResourceIntegrationVerification - a set of verification rules to query and check test related data
+type ResourceIntegrationVerification struct {
+	Name           string
+	ForeignKeyName string
+	ExpectedValues []ExpectedValue
+	Filter         func(sq squirrel.SelectBuilder, res *ResourceIntegrationTestData) squirrel.SelectBuilder
+	Relations      []*ResourceIntegrationVerification
+}
+
+// ExpectedValue - describes the data that expected to be in database after fetch
+type ExpectedValue struct {
+	Count int                    // expected count of items
+	Data  map[string]interface{} // expected data of items
+}
+
+// IntegrationTest - creates resources using terraform, fetches them to db and compares with expected values
+func IntegrationTest(t *testing.T, providerCreator func() *provider.Provider, resource ResourceIntegrationTestData) {
+	t.Parallel()
+	workdir, err := copyTfFiles(resource.Table.Name)
+	if err != nil {
+		t.Fatal(err)
+	}
+	lookPath := os.Getenv("TF_EXEC_PATH")
+	if lookPath == "" {
+		lookPath = "terraform"
+	}
+	execPath, err := exec.LookPath(lookPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tf, err := tfexec.NewTerraform(workdir, execPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = enableTerraformLog(tf, workdir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// prepare terraform variables
+	hostname, err := os.Hostname()
+	if err != nil {
+		t.Fatal(err)
+	}
+	resource.Suffix = simplifyString(hostname)
+	resource.Prefix = simplifyString(resource.Table.Name)
+	testSuffix := fmt.Sprintf("test_suffix=%s", resource.Suffix)
+	testPrefix := fmt.Sprintf("test_prefix=%s", resource.Prefix)
+
+	ctx := context.Background()
```

Comment:
maybe allow for an env variable for timeout? so if we apply takes longer than X it will fail because of context timeout, and we will fail test and destroy w.e possible. This is to avoid hanging tests